drivers/modem/quectel-bg9x.c

```diff
@@ -0,0 +1,1118 @@
+		     modem_context_sprint_ip_addr(addr), dst_port);
+
+	/* Send out the command. */
+	ret = modem_cmd_send(&mctx.iface, &mctx.cmd_handler,
+					 NULL, 0U, buf,
+						 &mdata.sem_response, K_SECONDS(1));
+	if (ret < 0) {
+		LOG_ERR("%s ret:%d", log_strdup(buf), ret);
+		errno = -ret;
+		return -1;
+	}
+
+	/* set command handlers */
+	ret = modem_cmd_handler_update_cmds(&mdata.cmd_handler_data,
+									cmd, ARRAY_SIZE(cmd), true);
+	if (ret < 0)
+		goto exit;
+
+	/* Wait for QI+OPEN */
+	ret = k_sem_take(&mdata.sem_sock_conn, MDM_CMD_CONN_TIMEOUT);
+	if (ret < 0) {
+		/* Didn't get the data prompt sadly. Need to exit. */
+		LOG_DBG("Timeout waiting for socket open");
+		goto exit;
+	}
+
+	ret = modem_cmd_handler_get_error(&mdata.cmd_handler_data);
+	if (ret != 0)
+		goto exit;
+
+	/* Connected successfully. */
+	sock->is_connected = true;
+	errno = 0;
+	return 0;
+
+exit:
+	(void) modem_cmd_handler_update_cmds(&mdata.cmd_handler_data,
+						NULL, 0U, false);
+	errno = -ret;
+	return -1;
+}
+
+/* Func: offload_close
+ * Desc: This function closes the connection with the remote client and
+ * frees the socket.
+ */
+static int offload_close(void *obj)
+{
+	struct modem_socket *sock = (struct modem_socket *) obj;
+	char		    buf[32] = {0};
+	int		    ret;
+
+	/* Make sure we assigned an id */
+	if (sock->id < mdata.socket_config.base_socket_num)
+		return 0;
+
+	/* Close the socket only if it is connected. */
+	if (sock->is_connected) {
+
+		snprintk(buf, sizeof(buf), "AT+QICLOSE=%d", sock->sock_fd);
+
+		/* Tell the modem to close the socket. */
+		ret = modem_cmd_send(&mctx.iface, &mctx.cmd_handler,
+						 NULL, 0U, buf,
+							 &mdata.sem_response, MDM_CMD_TIMEOUT);
+		if (ret < 0)
+			LOG_ERR("%s ret:%d", log_strdup(buf), ret);
+	}
+
+	modem_socket_put(&mdata.socket_config, sock->sock_fd);
+	return 0;
+}
+
+/* Func: offload_sendmsg
+ * Desc: This function sends messages to the modem.
+ */
+static ssize_t offload_sendmsg(void *obj, const struct msghdr *msg, int flags)
+{
+	ssize_t sent = 0;
+	int rc;
+
+	LOG_DBG("msg_iovlen:%d flags:%d", msg->msg_iovlen, flags);
+
+	for (int i = 0; i < msg->msg_iovlen; i++) {
+		const char *buf = msg->msg_iov[i].iov_base;
+		size_t len	= msg->msg_iov[i].iov_len;
+
+		while (len > 0) {
+			rc = offload_sendto(obj, buf, len, flags,
+							    msg->msg_name, msg->msg_namelen);
+			if (rc < 0) {
+				if (rc == -EAGAIN) {
+					k_sleep(MDM_SENDMSG_SLEEP);
+				} else {
+					sent = rc;
+					break;
+				}
+			} else {
+				sent += rc;
+				buf += rc;
+				len -= rc;
+			}
+		}
+	}
+
+	return (ssize_t) sent;
+}
+
+/* Func: modem_rx
+ * Desc: Thread to process all messages received from the Modem.
+ */
+static void modem_rx(void)
+{
+	while (true) {
+
+		/* Wait for incoming data */
+		k_sem_take(&mdata.iface_data.rx_sem, K_FOREVER);
+
+		/* process, then yield. */
+		mctx.cmd_handler.process(&mctx.cmd_handler, &mctx.iface);
+		k_yield();
```

Comment:
k_sem_take will re-sched the thread if I am not wrong. Does this need to yield explicit?

Reply:
Yup you're right. This was a copy-paste from the Sara-R4 driver, and I'm not sure why that driver resorted to an explicit yield.